programs/host/frontend.c

```diff
@@ -0,0 +1,957 @@
+                int32_t offset = 0;
+                int32_t whence = 0;
+                FILE * file = NULL;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                offset = item_uint32( inputs[0] );
+                whence = item_uint32( inputs[1] );
+                file_id = item_uint32( inputs[2] );
+                file = (FILE *)get_file_from_id( file_id );
+
+                if ( file )
+                {
+                    ret = 0;
+
+                    /* map whence with std lib */
+                    switch ( whence )
+                    {
+                        case MBEDTLS_SERIALIZE_FSEEK_SET:
+                            whence = SEEK_SET;
+                            break;
+                        case MBEDTLS_SERIALIZE_FSEEK_CUR:
+                            whence = SEEK_CUR;
+                            break;
+                        case MBEDTLS_SERIALIZE_FSEEK_END:
+                            whence = SEEK_END;
+                            break;
+                        default:
+                            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                            break;
+                    }
+
+                    if ( ret == 0 )
+                        ret = fseek( file, offset, whence );
+                }
+            }
+            break;
+        case MBEDTLS_SERIALIZE_FUNCTION_FTELL:
+            {
+                FILE * file = NULL;
+                int32_t file_id;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                file_id = item_uint32( inputs[0] );
+                file = (FILE *)get_file_from_id( file_id );
+
+                if ( file )
+                {
+                    ret = ftell( file );
+
+                    if ( ret >= 0 )
+                    {
+                        ALLOC_OUTPUT( 0, sizeof( int32_t ) );
+                        set_item_uint32( outputs[0], ret );
+                        ret = 0;
+                    }
+                }
+            }
+            break;
+        case MBEDTLS_SERIALIZE_FUNCTION_FERROR:
+            {
+                FILE * file = NULL;
+                int32_t file_id;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                file_id = item_uint32( inputs[0] );
+                file = (FILE *)get_file_from_id( file_id );
+
+                if ( file )
+                {
+                    ret = ferror( file );
+                }
+            }
+            break;
+        case MBEDTLS_SERIALIZE_FUNCTION_DOPEN:
+            {
+                char * path = NULL;
+                int32_t file_id;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                path = item_buffer( inputs[0] );
+                DBG( "open dir [%s]", path);
+                file_id = alloc_file_context();
+                DBG( "allocated dir id [%d]", file_id);
+                if ( file_id != -1 )
+                {
+                    DIR * dir = opendir( path );
```

Comment:
We need to call `mbedtls_opendir` here and not the Unix/POSIX function. This program should work on Windows without requiring a Unix compatibility layer. The same goes for the other directory-related functions.

Reply:
same comment as https://github.com/ARMmbed/mbedtls/pull/1029#discussion_r144245017